### PR TITLE
Apply entity equality to query filters

### LINKED ISSUE #18158: Simple QueryFilter breaks Linq-to-sql queries in EntityFrameworkCore
Reference to https://github.com/aspnet/EntityFrameworkCore/issues/13517
https://github.com/MintPlayer/QueryFilterIssue

**Migration from EntityFrameworkCore 2.2 to 3.0**
I'm using a QueryFilter in my DbContext to filter out the soft-deleted entities.

```
internal class MyDbContext : DbContext
{
	public DbSet<Person> People { get; set; }
	public DbSet<User> Users { get; set; }

	public MyDbContext() : base()
	{
	}

	public MyDbContext(DbContextOptions options) : base(options)
	{
	}

	protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
	{
		base.OnConfiguring(optionsBuilder);
		var connection_string = @"Server=(localdb)\mssqllocaldb;Database=QueryFilterIssue30;Trusted_Connection=True;ConnectRetryCount=0";
		optionsBuilder.UseSqlServer(connection_string);
	}

	protected override void OnModelCreating(ModelBuilder modelBuilder)
	{
		base.OnModelCreating(modelBuilder);

		// The following line breaks the app
		modelBuilder.Entity<Person>().HasQueryFilter(p => p.UserDelete == null);
	}
}
```

- EFCore 2.2: This code works
- EFCore 3.0: This code isn't working because of the QueryFilter.

However, imagine following QueryFilter:
`modelBuilder.Entity<Person>().HasQueryFilter(p => (p.FirstName != "") & (p.LastName != ""));`
This code is in fact working.

Suppose the QueryFilters applied to the DbSet are also included in the Linq-to-sql conversion, and not being run client-side, this code should be executed without a problem.

### Summary:
It appears that since EF Core 3.0 applying a QueryFilter containing a relationship condition breaks EF.

- Applying a QueryFilter based on fields works perfectly fine
- Removing the QueryFilter and moving the statement to the LINQ query explicitly works as expected
- Running the same code in EFCore 2.2 works.
- You can even combine AspNetCore 3.0 and EFCore 2.2

### Steps to reproduce

- git clone https://github.com/MintPlayer/QueryFilterIssue.git
- cd QueryFilterIssue/DotNetCore22/QueryFilterIssue22.Data
- dotnet ef database update
- cd QueryFilterIssue/DotNetCore30/QueryFilterIssue30.Data
- dotnet ef database update
- Add data to the databases
- Set **QueryFilterIssue22.Web** as startup project and run
- The people are successfully fetched from the database
- Set **QueryFilter30.Web** as startup project and run
- Notice the following error

> InvalidOperationException: The LINQ expression 'Where<TransparentIdentifier<Person, User>>(
> source: LeftJoin<Person, User, Nullable<int>, TransparentIdentifier<Person, User>>(
> outer: DbSet<Person>,
> inner: DbSet<User>,
> outerKeySelector: (p) => Property<Nullable<int>>(p, "UserDeleteId"),
> innerKeySelector: (u) => Property<Nullable<int>>(u, "Id"),
> resultSelector: (o, i) => new TransparentIdentifier<Person, User>(
> Outer = o,
> Inner = i
> )),
> predicate: (p) => p.Inner == (Unhandled parameter: __ef_filter__p_0))' could not be translated. Either rewrite the query in a form that can be translated, or switch to client evaluation explicitly by inserting a call to either AsEnumerable(), AsAsyncEnumerable(), ToList(), or ToListAsync(). See https://go.microsoft.com/fwlink/?linkid=2101038 for more information.

- Disable the QueryFilter in `MyDbContext`. Optionally enable the other QueryFilter.
- Run the project again. Notice the working app.
- Optionally enable the Where clause in the PersonRepository.
- Run the project again. Notice the working app.

### Further technical details

EF Core version: 2.2.0 versus 3.0.0
Database provider: Microsoft.EntityFrameworkCore.SqlServer 2.2.0 versus 3.0.0
Target framework: .NET Core 2.2 versus 3.0
Operating system: Windows 10
IDE: Visual Studio 2019 16.3.1

### PR DESCRIPTION
**Pull request by roji**
Fixes #18158

Added the entity equality scenario to an existing query filter in Northwind - not super pretty but works. I can do a QueryBugTest instead (extra db), let me know.

### CODE REVIEW

File: src/EFCore/Query/Internal/EntityEqualityRewritingExpressionVisitor.cs [hunk 1]
```diff
@@ -709,12 +709,21 @@ protected virtual Expression VisitOfType(MethodCallExpression methodCallExpressi
             return new EntityReferenceExpression(updatedMethodCall, castEntityType);
         }
 
+        public virtual LambdaExpression RewriteSingleParamLambda(LambdaExpression lambda, IEntityType entityType)
```
[the remark this answers is not in the document]
Comment (roji) [reply src/EFCore/Query/Internal/EntityEqualityRewritingExpressionVisitor.cs]: The visitor is internal.

The point is that unlike the other visitors invoked by nav expander, here an additional IEntityType needs to be provided, so we need another entry point into the EE visitor. Any suggestions?

[the remark this answers is not in the document]
Comment (smitpatel) [reply src/EFCore/Query/Internal/EntityEqualityRewritingExpressionVisitor.cs]: use the predicate create `dbset.Where()` and pass into entity equality using `Rewrite` method how it is done from outside perspective already.

[the remark this answers is not in the document]
Comment (roji) [reply src/EFCore/Query/Internal/EntityEqualityRewritingExpressionVisitor.cs]: Thanks for the suggestion. How's this new version?

File: src/EFCore/Query/Internal/NavigationExpandingExpressionVisitor.cs [hunk 2]
```diff
@@ -172,13 +175,23 @@ private Expression ApplyQueryFilter(NavigationExpansionExpression navigationExpa
                         filterPredicate = queryFilter;
                         filterPredicate = (LambdaExpression)_parameterExtractingExpressionVisitor.ExtractParameters(filterPredicate);
                         filterPredicate = (LambdaExpression)_enumerableToQueryableMethodConvertingExpressionVisitor.Visit(filterPredicate);
+
+                        // We need to do entity equality, but that requires a full method call on a query root to properly flow the
+                        // entity information through. Construct a MethodCall wrapper for the predicate with the proper query root.
+                        var queryableType = typeof(EntityQueryable<>).MakeGenericType(sequenceType);
```
Comment (smitpatel) [inline src/EFCore/Query/Internal/NavigationExpandingExpressionVisitor.cs]: Search for `CreateEntityQueryableExpression`

File: src/EFCore/Query/Internal/EntityEqualityRewritingExpressionVisitor.cs [hunk 3]
```diff
@@ -319,7 +319,7 @@ protected override Expression VisitMethodCall(MethodCallExpression methodCallExp
                 {
                     // Visit all arguments, rewriting the single lambda to replace its parameter expression
                     newArguments[i] = arguments[i].GetLambdaOrNull() is LambdaExpression lambda
-                        ? Unwrap(RewriteAndVisitLambda(lambda, newSourceWrapper))
+                        ? Unwrap(RewriteSingleParamLambda(lambda, newSourceWrapper))
```
Comment (smitpatel) [inline src/EFCore/Query/Internal/EntityEqualityRewritingExpressionVisitor.cs]: Revert changes in this file.

File: src/EFCore/Query/Internal/EntityEqualityRewritingExpressionVisitor.cs [hunk 4]
```diff
@@ -723,16 +729,22 @@ protected virtual LambdaExpression RewriteAndVisitLambda(LambdaExpression lambda
...
                 lambda.TailCall,
                 lambda.Parameters);
+        }
 
         /// <summary>
         ///     Replaces the lambda's two parameters with type wrappers based on the given sources, and then visits
         ///     the lambda's body.
         /// </summary>
-        protected virtual LambdaExpression RewriteAndVisitLambda(
+        protected virtual LambdaExpression RewriteTwoParamLambda(
```
Comment (smitpatel) [inline src/EFCore/Query/Internal/EntityEqualityRewritingExpressionVisitor.cs]: Also make all the methods inside this class private unless overriding Visit* methods & public entry point. We should not be making anything protected till we decide to move this class outside of .Internal namespace.

File: src/EFCore/Query/Internal/NavigationExpandingExpressionVisitor.cs (hunk 2 above)
Comment (roji) [reply src/EFCore/Query/Internal/NavigationExpandingExpressionVisitor.cs]: Great, thanks! Applied.

File: src/EFCore/Query/Internal/EntityEqualityRewritingExpressionVisitor.cs (hunk 3 above)
Comment (roji) [reply src/EFCore/Query/Internal/EntityEqualityRewritingExpressionVisitor.cs]: Done

File: src/EFCore/Query/Internal/EntityEqualityRewritingExpressionVisitor.cs (hunk 4 above)
Comment (roji) [reply src/EFCore/Query/Internal/EntityEqualityRewritingExpressionVisitor.cs]: Was not aware of this rule - it makes it somewhat useless to make the class pubternal in the first place.

Done.

File: src/EFCore/Query/Internal/NavigationExpandingExpressionVisitor.cs (hunk 2 above)
Comment (smitpatel) [inline src/EFCore/Query/Internal/NavigationExpandingExpressionVisitor.cs]: unused variable.

**Comment by roji** [comment]
@smitpatel note tiny change to support inheritance scenarios... Am merging as everything is green now.

**Pull request merged**